# kmeans Clustering with Scikit-Learn Assignment

In this assignment we will be using scikit-learn to identify client segemnts in a database of customers from a wholesale retailer.

The original dataset (which can be found here) contains the follow features that we will use for clustering.

FRESH: annual spending (m.u.) on fresh products (Continuous);
MILK: annual spending (m.u.) on milk products (Continuous);
GROCERY: annual spending (m.u.)on grocery products (Continuous);
FROZEN: annual spending (m.u.)on frozen products (Continuous)
DETERGENTS_PAPER: annual spending (m.u.) on detergents and paper products (Continuous)
DELICATESSEN: annual spending (m.u.)on and delicatessen products (Continuous);

First, we need to load the appropriate libraries for cleaning the data and then running a kMeans clustering algorithm. Again, we also set a random seed for replication.

In [1]:
from sklearn import cluster
import pandas as pd
import numpy as np
np.random.seed(100)

Here we use the read_csv function from pandas to read in our dataset and take a subset of variables used for training.

In [2]:
df = pd.read_csv('Wholesale customers data.csv')
df = df[['Fresh','Milk','Grocery','Frozen','Detergents_Paper','Delicassen']]

Next, we use the KMeans function to group the data into three natural clusters. We can then print out the assigned labels using .labels_

In [3]:
k_means = cluster.KMeans(n_clusters=3)
k_means.fit(df) 
k_means.labels_

array([0, 0, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       2, 1, 0, 0, 0, 2, 1, 0, 0, 0, 1, 0, 0, 1, 0, 2, 1, 1, 0, 0, 2, 0, 2,
       2, 2, 0, 2, 0, 0, 1, 0, 1, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 2, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 2, 2, 1, 0, 1, 0, 0,
       2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0,
       0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 1, 0, 0, 2, 0,
       0, 0, 2, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0,
       1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 2, 2, 0, 2, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0,

Here we take a subset of the observations in our wholesale customer database that were assigned to cluster 0 and print out the average spent per category of purchases as well as the total. We can inspect the observations in cluster 0 by using: df.loc[i,]

In [4]:
i, = np.where(k_means.labels_==0)
print('Mean:')
print(df.loc[i,].mean())
print('\nTotal:')
print(df.loc[i,].mean().sum()/6)

df.loc[i,]

Mean:
Fresh               8253.469697
Milk                3824.603030
Grocery             5280.454545
Frozen              2572.660606
Detergents_Paper    1773.057576
Delicassen          1137.496970
dtype: float64

Total:
3806.95707071


Fresh   Milk  Grocery  Frozen  Detergents_Paper  Delicassen
0    12669   9656     7561     214              2674        1338
1     7057   9810     9568    1762              3293        1776
2     6353   8808     7684    2405              3516        7844
3    13265   1196     4221    6404               507        1788
5     9413   8259     5126     666              1795        1451
6    12126   3199     6975     480              3140         545
7     7579   4956     9426    1669              3321        2566
8     5963   3648     6192     425              1716         750
10    3366   5403    12974    4400              5977        1744
11   13146   1124     4523    1420               549         497
13   21217   6208    14982    3095              6707         602
15   10253   1114     3821     397               964         412
16    1020   8816    12121     134              4508        1080
17    5876   6157     2933     839               370        4478
18   18601   6327    10099    2205              2767        3181
19    7780   2495     9464     669              2518         501
20   17546   4519     4602    1066              2259        2124
21    5567    871     2010    3383               375         569
25   16165   4230     7595     201              4003          57
26    9898    961     2861    3151               242         833
27   14276    803     3045     485               100         518
30   18815   3610    11107    1148              2134        2963
31    2612   4339     3133    2088               820         985
32   21632   1318     2886     266               918         405
34    1502   1979     2262     425               483         395
35     688   5491    11091     833              4239         436
37   15168  10556    12477    1920              6506         714
41   19176   3065     5956    2033              2575        2802
42   10850   7555    14961     188              6899          46
44    9670   7027    10471     541              4618          65
..     ...    ...      ...     ...               ...         ...
399   9612    577      935    1601               469         375
400   4446    906     1238    3576               153        1014
404  18073   4613     3444    4324               914         715
405   6884   1046     1167    2069               593         378
408   8257   3880     6407    1646              2730         344
409   8708   3634     6100    2349              2123        5137
410   6633   2096     4563    1389              1860        1892
411   2126   3289     3281    1535               235        4365
412     97   3605    12400      98              2970          62
413   4983   4859     6633   17866               912        2435
414   5969   1990     3417    5679              1135         290
415   7842   6046     8552    1691              3540        1874
416   4389  10940    10908     848              6728         993
417   5065   5499    11055     364              3485        1063
419   8861   3783     2223     633              1580        1521
420   4456   5266    13227      25              6818        1393
421  17063   4847     9053    1031              3415        1784
423  17565   3686     4657    1059              1803         668
424  16980   2884    12232     874              3213         249
425  11243   2408     2593   15348               108        1886
426  13134   9347    14316    3141              5079        1894
428   3047   5970     4910    2198               850         317
429   8607   1750     3580      47                84        2501
430   3097   4230    16483     575               241        2080
431   8533   5506     5160   13486              1377        1498
432  21117   1162     4754     269              1328         395
433   1982   3218     1493    1541               356        1449
434  16731   3922     7994     688              2371         838
438  10290   1981     2232    1038               168        2125
439   2787   1698     2510      65      

In [5]:
i, = np.where(k_means.labels_==1)
print('Mean:')
print(df.loc[i,].mean())
print('\nTotal:')
print(df.loc[i,].mean().sum()/6)

df.loc[i,]

Mean:
Fresh               35941.400000
Milk                 6044.450000
Grocery              6288.616667
Frozen               6713.966667
Detergents_Paper     1039.666667
Delicassen           3049.466667
dtype: float64

Total:
9846.26111111


Fresh   Milk  Grocery  Frozen  Detergents_Paper  Delicassen
4     22615   5410     7198    3915              1777        5185
12    31714  12319    11757     287              3881        2931
14    24653   9465    12091     294              5058        2168
22    31276   1917     4469    9408              2381        4334
24    22647   9776    13792    2915              4482        5778
29    43088   2100     2609    1200              1107         823
33    29729   4786     7326    6130               361        1083
36    29955   4362     5428    1729               862        4626
39    56159    555      902   10002               212        2916
40    24025   4332     4757    9510              1145        5864
52    40721   3916     5876     532              2587        1278
54    27329   1449     1947    2436               204        1333
87    43265   5025     8117    6312              1579       14351
89    24904   3836     5330    3443               454        3178
103   56082   3504     8906   18028              1480        2498
124   36050   1642     2961    4787               500        1621
125   76237   3473     7102   16538               778         918
129   42312    926     1510    1718               410        1819
141   30379  13252     5189     321                51        1450
142   37036   7152     8253    2995                20           3
149   31812   1433     1651     800               113        1440
176   45640   6958     6536    7368              1532         230
181  112151  29627    18148   16745              4948        8550
183   36847  43950    20170   36534               239       47943
196   30624   7209     4897   18711               763        2876
202   25203  11487     9490    5065               284        6854
232   25962   1780     3838     638               284         834
239   47493   2567     3779    5243               828        2253
240   22096   3575     7041   11422               343        2564
241   24929   1801     2475    2216               412        1047
253   29526   7961    16966     432               363        1391
255   31614    489     1495    3242               111         615
258   56083   4563     2124    6422               730        3321
259   53205   4959     7336    3012               967         818
262   23257   1372     1677     982               429         655
273   36817   3045     1493    4802               210        1824
276   27901   3749     6964    4479               603        2503
282   49063   3965     4252    5970              1041        1404
283   25767   3613     2013   10303               314        1384
284   68951   4411    12609    8692               751        2406
285   40254    640     3600    1042               436          18
289   42786    286      471    1388                32          22
311   29635   2335     8280    3046               371         117
324   25977   3587     2464    2369               140        1092
325   32717  16784    13626   60869              1272        5609
335   27082   6817    10790    1365              4111        2139
347   27380   7184    12311    2809              4621        1022
370   39679   3944     4955    1364               523        2235
377   38793   3154     2648    1034                96        1242
380   28257    944     2146    3881               600         270
382   34454   7435     8469    2540              1711        2893
393   23632   6730     3842    8620               385         819
401   27167   2801     2128   13223                92        1902
402   26539   4753     5091     220                10         340
403   25606  11006     4604     127               632         288
406   25066   5010     5026    9806              1092         960
422   26400   1377     4172     830               948        1218
427   31012  16687     5429   15082               439        1163
435   29703  12051    16027   13135               182        2204
436   39228   1431      764    4510           

In [6]:
i, = np.where(k_means.labels_==2)
print('Mean:')
print(df.loc[i,].mean())
print('\nTotal:')
print(df.loc[i,].mean().sum()/6)

df.loc[i,]

Mean:
Fresh                8000.04
Milk                18511.42
Grocery             27573.90
Frozen               1996.68
Detergents_Paper    12407.36
Delicassen           2252.02
dtype: float64

Total:
11790.2366667


Fresh   Milk  Grocery  Frozen  Detergents_Paper  Delicassen
9     6006  11093    18881    1159              7425        2098
23   26373  36423    22019    5154              4337       16523
28    4113  20484    25957    1158              8604        5206
38    4591  15729    16709      33              6956         433
43     630  11095    23998     787              9529          72
45    5181  22044    21531    1740              7353        4985
46    3103  14069    21955    1668              6792        1452
47   44466  54259    55571    7782             24171        6465
49    4967  21412    28921    1798             13583        1163
56    4098  29892    26866    2616             17740        1340
61   35942  38369    59598    3254             26701        2017
65      85  20959    45828      36             24231        1423
77   12205  12697    28540     869             12034        1009
85   16117  46197    92780    1026             40827        2944
86   22925  73498    32114     987             20070         903
92    9198  27472    32034    3232             18906        5130
109   1406  16729    28986     673               836           3
111  12579  11114    17569     805              6457        1519
145  22039   8384    34792      42             12591        4430
155   1989  10690    19460     233             11577        2153
163   5531  15726    26870    2367             13726         446
165  15615  12653    19858    4425              7108        2379
171    200  25862    19816     651              8773        6250
173    514   7677    19805     937              9836         716
200   3067  13240    23127    3941              9959         731
201   4484  14399    24708    3549             14235        1681
205   1107  11711    23596     955              9265         710
209   5550  12729    16767     864             12420         797
211  12119  28326    39694    4736             19410        2870
216   2532  16599    36486     179             13308         674
251   6134  23133    33586    6746             18594        5121
265   5909  23527    13699   10155               830        3636
266    572   9763    22182    2221              4882        2563
268  11908   8053    19847    1069              6374         698
301   5283  13316    20399    1809              8752         172
304    161   7460    24773     617             11783        2410
306   6468  12867    21570    1840              7558        1543
309    918  20655    13567    1465              6846         806
312   2137   3737    19172    1274             17120         142
319   9759  25071    17645    1128             12408        1625
331  11223  14881    26839    1234              9606        1102
333   8565   4980    67298     131             38102        1215
343   1689   6964    26316    1456             15469          37
349   5981  14641    20521    2005             12218         445
351   1210  10044    22294    1741             12638        3137
353    117   6264    21203     228              8682        1111
384  10683  21858    15400    3635               282        5120
407   7362  12844    18683    2854              7883         553
418    660   8494    18622     133              6740         776
437  14531  15488    30243     437             14841        1867

# It looks like the clusters indicate that there are three customer segments: small (~3,800), medium ( ~9,845 ), and high ( ~11,790) spending groups.